File: word-count-beam/src/main/java/org/apache/beam/examples/subprocess/utils/FileUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.beam.examples.subprocess.utils;

import static java.nio.charset.StandardCharsets.UTF_8;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.ReadableByteChannel;
import java.nio.channels.WritableByteChannel;
import java.nio.file.FileAlreadyExistsException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import org.apache.beam.examples.subprocess.configuration.SubProcessConfiguration;
import org.apache.beam.sdk.io.FileSystems;
import org.apache.beam.sdk.io.fs.ResolveOptions.StandardResolveOptions;
import org.apache.beam.sdk.io.fs.ResourceId;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Utilities for dealing with movement of files from object stores and workers. */
@SuppressWarnings({
  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
})
public class FileUtils {

  private static final Logger LOG = LoggerFactory.getLogger(FileUtils.class);

  public static ResourceId getFileResourceId(String directory, String fileName) {
    ResourceId resourceID = FileSystems.matchNewResource(directory, true);
    return resourceID.getCurrentDirectory().resolve(fileName, StandardResolveOptions.RESOLVE_FILE);
  }

  public static String toStringParams(ProcessBuilder builder) {
    return String.join(",", builder.command());
  }

  public static String copyFileFromWorkerToGCS(
      SubProcessConfiguration configuration, Path fileToUpload) throws Exception {

    Path fileName;

    if ((fileName = fileToUpload.getFileName()) == null) {
      throw new IllegalArgumentException("FileName can not be null.");
    }

    ResourceId sourceFile = getFileResourceId(configuration.getWorkerPath(), fileName.toString());

    LOG.info("Copying file from worker " + sourceFile);

    ResourceId destinationFile =
        getFileResourceId(configuration.getSourcePath(), fileName.toString());
    // TODO currently not supported with different schemas for example GCS to local, else could use
    // FileSystems.copy(ImmutableList.of(sourceFile), ImmutableList.of(destinationFile));
    try {
      return copyFile(sourceFile, destinationFile);
    } catch (Exception ex) {
      LOG.error(
          String.format("Error copying file from %s  to %s", sourceFile, destinationFile), ex);
      throw ex;
    }
  }

  public static String copyFileFromGCSToWorker(ExecutableFile execuableFile) throws Exception {

    ResourceId sourceFile =
        FileSystems.matchNewResource(execuableFile.getSourceGCSLocation(), false);
    ResourceId destinationFile =
        FileSystems.matchNewResource(execuableFile.getDestinationLocation(), false);
    try {
      LOG.info(
          String.format(
              "Moving File %s to %s ",
              execuableFile.getSourceGCSLocation(), execuableFile.getDestinationLocation()));
      Path path = Paths.get(execuableFile.getDestinationLocation());

      if (path.toFile().exists()) {
        LOG.warn(
            String.format(
                "Overwriting file %s, should only see this once per worker.",
                execuableFile.getDestinationLocation()));
      }
      copyFile(sourceFile, destinationFile);
      path.toFile().setExecutable(true);
      return path.toString();

    } catch (Exception ex) {
      LOG.error(String.format("Error moving file : %s ", execuableFile.fileName), ex);
      throw ex;
    }
  }

  public static String copyFile(ResourceId sourceFile, ResourceId destinationFile)
      throws IOException {

    try (WritableByteChannel writeChannel = FileSystems.create(destinationFile, "text/plain")) {
      try (ReadableByteChannel readChannel = FileSystems.open(sourceFile)) {

        final ByteBuffer buffer = ByteBuffer.allocateDirect(16 * 1024);
        while (readChannel.read(buffer) != -1) {
          buffer.flip();
          writeChannel.write(buffer);
          buffer.compact();
        }
        buffer.flip();
        while (buffer.hasRemaining()) {
          writeChannel.write(buffer);
        }
      }
    }

    return destinationFile.toString();
  }

  /**
   * Create directories needed based on configuration.
   *
   * @param configuration
   * @throws IOException
   */
  public static void createDirectoriesOnWorker(SubProcessConfiguration configuration)
      throws IOException {

    try {

      Path path = Paths.get(configuration.getWorkerPath());

      if (!path.toFile().exists()) {
        Files.createDirectories(path);
        LOG.info(String.format("Created Folder %s ", path.toFile()));
      }
    } catch (FileAlreadyExistsException ex) {
      LOG.warn(
          String.format(
              " Tried to create folder %s which already existsed, this should not happen!",
              configuration.getWorkerPath()),
          ex);
    }
  }

  public static String readLineOfLogFile(Path path) {

    try (BufferedReader br = Files.newBufferedReader(Paths.get(path.toString()), UTF_8)) {
      return br.readLine();
    } catch (IOException e) {
      LOG.error("Error reading the first line of file", e);
    }

    // `return empty string rather than NULL string as this data is often used in further logging
    return "";
  }
}

File: playground/backend/internal/utils/file_utils.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package utils

import (
	"errors"
	"path/filepath"
	"strings"

	pb "beam.apache.org/playground/backend/internal/api/v1"
	"beam.apache.org/playground/backend/internal/logger"
)

const (
	javaMainMethod        = "public static void main(String[] args)"
	goMainMethod          = "func main()"
	pythonMainMethod      = "if __name__ == '__main__'"
	scioMainMethod        = "def main(cmdlineArgs: Array[String])"
	defaultJavaFileName   = "main.java"
	defaultGoFileName     = "main.go"
	defaultPythonFileName = "main.py"
	defaultScioFileName   = "main.scala"
	javaExt               = ".java"
	classExt              = ".class"
	goExt                 = ".go"
	pythonExt             = ".py"
	scioExt               = ".scala"

	mismatchExtAndSDKErrMsg = "file extension doesn't match to the specified SDK"
)

// GetFileName returns the valid file name.
func GetFileName(name, content string, sdk pb.Sdk) (string, error) {
	if name == "" {
		logger.Warn("The name of the file is empty. Will be used default value")
		switch sdk {
		case pb.Sdk_SDK_JAVA:
			return defaultJavaFileName, nil
		case pb.Sdk_SDK_GO:
			return defaultGoFileName, nil
		case pb.Sdk_SDK_PYTHON:
			return defaultPythonFileName, nil
		case pb.Sdk_SDK_SCIO:
			return defaultScioFileName, nil
		}
	}
	return getCorrectFileName(name, content, sdk)
}

// getCorrectFileName returns the correct file name.
func getCorrectFileName(name, content string, sdk pb.Sdk) (string, error) {
	ext := getExtOrExtBasedOnContent(filepath.Ext(name), content)
	if ext != "" && !isValidFileExtensionAndSDK(ext, sdk) {
		return "", errors.New(mismatchExtAndSDKErrMsg)
	}
	switch sdk {
	case pb.Sdk_SDK_JAVA:
		return getCorrectNameOrDefault(ext, javaExt, defaultJavaFileName, name), nil
	case pb.Sdk_SDK_GO:
		return getCorrectNameOrDefault(ext, goExt, defaultGoFileName, name), nil
	case pb.Sdk_SDK_PYTHON:
		return getCorrectNameOrDefault(ext, pythonExt, defaultPythonFileName, name), nil
	case pb.Sdk_SDK_SCIO:
		return getCorrectNameOrDefault(ext, scioExt, defaultScioFileName, name), nil
	default:
		return name, nil
	}
}

// isValidFileExtensionAndSDK returns a flag indicating the result of validation
func isValidFileExtensionAndSDK(ext string, sdk pb.Sdk) bool {
	switch ext {
	case javaExt:
		return sdk == pb.Sdk_SDK_JAVA
	case goExt:
		return sdk == pb.Sdk_SDK_GO
	case pythonExt:
		return sdk == pb.Sdk_SDK_PYTHON
	case scioExt:
		return sdk == pb.Sdk_SDK_SCIO
	default:
		return false
	}
}

// getExtBasedOnContent return a file extension
func getExtOrExtBasedOnContent(ext, content string) string {
	if ext == "" {
		return getExtBasedOnContent(content)
	}
	return ext
}

// getExtBasedOnContent return a file extension based on the content
func getExtBasedOnContent(content string) string {
	if strings.Contains(content, javaMainMethod) {
		return javaExt
	}
	if strings.Contains(content, goMainMethod) {
		return goExt
	}
	if strings.Contains(content, pythonMainMethod) {
		return pythonExt
	}
	if strings.Contains(content, scioMainMethod) {
		return scioExt
	}
	return ""
}

// getCorrectNameOrDefault returns the correct file name or default name.
func getCorrectNameOrDefault(actualExt, correctExt, defaultFileName, name string) string {
	if actualExt == "" {
		logger.Infof("The name of the file does not have extension. Default value (%s) will be used", correctExt)
		if name == "" {
			return defaultFileName
		}
		return name + correctExt
	}
	if actualExt != correctExt {
		logger.Infof("The name of the file has wrong extension. Correct extension (%s) will be used according to sdk", correctExt)
		return name[0:len(name)-len(actualExt)] + correctExt
	}
	if filepath.Ext(name) == "" {
		return name + correctExt
	}
	return name
}

// IsFileMain returns true if the file content has a main function, otherwise false.
func IsFileMain(content string, sdk pb.Sdk) bool {
	switch sdk {
	case pb.Sdk_SDK_JAVA:
		return strings.Contains(content, javaMainMethod)
	case pb.Sdk_SDK_GO:
		return strings.Contains(content, goMainMethod)
	case pb.Sdk_SDK_PYTHON:
		return strings.Contains(content, pythonMainMethod)
	case pb.Sdk_SDK_SCIO:
		return strings.Contains(content, scioMainMethod)
	default:
		return false
	}
}

// ToSDKFromExt returns SDK according to a specified extension.
func ToSDKFromExt(ext string) pb.Sdk {
	switch ext {
	case javaExt, classExt:
		return pb.Sdk_SDK_JAVA
	case goExt:
		return pb.Sdk_SDK_GO
	case scioExt:
		return pb.Sdk_SDK_SCIO
	case pythonExt:
		return pb.Sdk_SDK_PYTHON
	default:
		return pb.Sdk_SDK_UNSPECIFIED
	}
}

func TrimExtension(filename string) string {
	return strings.TrimSuffix(filename, filepath.Ext(filename))
}
